documentation/writing_plugins.md

```diff
@@ -1,151 +1,492 @@
 # Writing plugins
 
-Users can write plugins that live within a module.
+Plugins are written as tasks that are loaded from modules when you run Bolt. If
+you are familiar with writing tasks, then you can write plugins! If you are not
+familiar with how to write a task, we recommend that you read more about
+[writing tasks](writing_tasks.md) before continuing on.
 
-Include your own plugins within a module by adding a `bolt_plugin.json` file to
-the top level of that module. This file should describe the plugin's
-configuration and can optionally map specific hooks to tasks that implement
-them. In most cases those tasks will run on the default `localhost` target
-without any configuration from `inventory.yaml`. The only exception is the
-`puppet_library` hook, where the task will be run on the target that needs the
-Puppet library.
+## Supported languages
 
-## Configuration
+Because plugins are written as tasks, they can be written in any scripting
+language supported by the target it runs on. Plugins are always run on the
+`localhost` target without any configuration from the inventory file, so as long
+as your controller can interpret the scripting language the plugin is written
+with, you can run the plugin.
 
-Plugins can use configuration from the `bolt-project.yaml` or `bolt-defaults.yaml` files. To allow a
-plugin to be configured, add a `parameters` section to the [task
-metadata](writing_tasks.md#task-metadata).
+For Unix-like controllers, you should include a shebang line at the top of of
+the plugin file to ensure it is executed using the correct interpreter.
+
+```python
+#!/usr/bin/env python
+
+import json, os, sys
+
+...
+```
+
+## Module structure
+
+Because plugins are written as tasks, they are part of a module. A simple
+module including a plugin will have a structure similar to this:
```

Comment:
Ah, that should definitely be a `my_plugin` module directory with a `resolve_reference` plugin. I was originally trying to avoid introducing the concept of hooks too early, but it just adds more confusion.